src/Netscript/APIWrapper.ts

```diff
@@ -0,0 +1,157 @@
+  workerScript: WorkerScript;
+  function: string;
+  helper: WrappedNetscriptHelpers;
+};
+
+type NetscriptHelpers = {
+  updateDynamicRam: (fnName: string, ramCost: number) => void;
+  makeRuntimeErrorMsg: (caller: string, msg: string) => string;
+  string: (funcName: string, argName: string, v: unknown) => string;
+  number: (funcName: string, argName: string, v: unknown) => number;
+  boolean: (v: unknown) => boolean;
+  getServer: (hostname: string, callingFnName: string) => BaseServer;
+  checkSingularityAccess: (func: string) => void;
+  hack: (hostname: any, manual: any, { threads: requestedThreads, stock }?: any) => Promise<number>;
+  getValidPort: (funcName: string, port: any) => IPort;
+};
+
+type WrappedNetscriptHelpers = {
+  updateDynamicRam: (ramCost: number) => void;
+  makeRuntimeErrorMsg: (msg: string) => string;
+  string: (argName: string, v: unknown) => string;
+  number: (argName: string, v: unknown) => number;
+  boolean: (v: unknown) => boolean;
+  getServer: (hostname: string) => BaseServer;
+  checkSingularityAccess: () => void;
+  hack: (hostname: any, manual: any, { threads: requestedThreads, stock }?: any) => Promise<number>;
+  getValidPort: (port: any) => IPort;
+};
+
+function wrapFunction(
+  helpers: NetscriptHelpers,
+  wrappedAPI: any,
+  workerScript: WorkerScript,
+  func: (_ctx: NetscriptContext) => (...args: unknown[]) => unknown,
+  ...tree: string[]
+): void {
+  const functionPath = tree.join(".");
+  const functionName = tree.pop();
+  if (typeof functionName !== "string") {
+    throw makeRuntimeRejectMsg(workerScript, "Failure occured while wrapping netscript api");
+  }
+  const ctx = {
+    makeRuntimeErrorMsg: (message: string) => {
+      return helpers.makeRuntimeErrorMsg(functionPath, message);
+    },
+    log: (message: () => string) => {
+      workerScript.log(functionPath, message);
+    },
+    workerScript,
+    function: functionName,
+    helper: {
+      updateDynamicRam: (ramCost: number) => helpers.updateDynamicRam(functionName, ramCost),
+      makeRuntimeErrorMsg: (msg: string) => helpers.makeRuntimeErrorMsg(functionPath, msg),
+      string: (argName: string, v: unknown) => helpers.string(functionPath, argName, v),
+      number: (argName: string, v: unknown) => helpers.number(functionPath, argName, v),
+      boolean: helpers.boolean,
+      getServer: (hostname: string) => helpers.getServer(hostname, functionPath),
+      checkSingularityAccess: () => helpers.checkSingularityAccess(functionName),
+      hack: helpers.hack,
+      getValidPort: (port: any) => helpers.getValidPort(functionPath, port),
+    },
+  };
+  function wrappedFunction(...args: unknown[]): unknown {
+    helpers.updateDynamicRam(ctx.function, getRamCost(Player, ...tree, ctx.function));
+    return func(ctx)(...args);
+  }
+  const parent = getNestedProperty(wrappedAPI, ...tree);
+  Object.defineProperty(parent, functionName, {
+    value: wrappedFunction,
+    writable: true,
```

Comment:
false probably, I didn't know about defineProperty, it's cool.

Reply:
`writable` has to be true actually or the other wrapper fails